src/lib/settings/GatewayStorage.js

```diff
@@ -0,0 +1,233 @@
+	 */
+	async init() {
+		if (this.ready) throw new Error('KlasaGatewayStorage already inited.');
+		await this.initSchema();
+		await this.initTable();
+
+		this.ready = true;
+	}
+
+	/**
+	 * Inits the table for its use in this gateway.
+	 * @since 0.5.0
+	 * @private
+	 */
+	async initTable() {
+		const hasTable = await this.provider.hasTable(this.type);
+		if (!hasTable) await this.provider.createTable(this.type, this.sql ? this.sqlSchema.map(([k, v]) => `${k} ${v}`).join(', ') : undefined);
+	}
+
+	/**
+	 * Inits the schema, creating a file if it does not exist, and returning the current schema or the default.
+	 * @since 0.5.0
+	 * @returns {Promise<Schema>}
+	 * @private
+	 */
+	async initSchema() {
+		await fs.ensureDir(this.baseDir);
+		const schema = await fs.readJSON(this.filePath)
+			.catch(() => fs.outputJSONAtomic(this.filePath, this.defaultSchema).then(() => this.defaultSchema));
+		this.schema = new Schema(this.client, this, schema, null, '');
+		return this.schema;
+	}
+
+	/**
+	 * Parses an entry
+	 * @since 0.5.0
+	 * @param {Object} entry An entry to parse.
+	 * @returns {Object}
+	 * @private
+	 */
+	parseEntry(entry) {
+		const object = {};
+		for (const piece of this.schema.getValues()) {
+			// If the key does not exist in the schema, ignore it.
+			if (typeof entry[piece.path] === 'undefined') continue;
+
+			if (piece.path.includes('.')) {
+				const path = piece.path.split('.');
+				let refObject = object;
+				for (let a = 0; a < path.length - 1; a++) {
+					const key = path[a];
+					if (typeof refObject[key] === 'undefined') refObject[path[a]] = {};
+					refObject = refObject[key];
+				}
+				refObject = GatewayStorage._parseSQLValue(entry[path[path.length - 1]], piece);
+			} else {
+				object[piece.path] = GatewayStorage._parseSQLValue(entry[piece.path], piece);
+			}
+		}
+
+		return object;
+	}
+
+	/**
+	 * Make an error that can or not have a valid Guild.
+	 * @since 0.5.0
+	 * @param {KlasaGuild} guild The guild to get the language from.
+	 * @param {(string|number)} code The code of the error.
+	 * @param {(string|Error)} error The error.
+	 * @returns {string}
+	 * @private
+	 * @static
+	 */
+	static throwError(guild, code, error) {
+		if (guild && guild.language && typeof guild.language.get === 'function') return guild.language.get(code);
+		return `ERROR: [${code}]: ${error}`;
+	}
+
+	/**
+	 * Parse SQL values.
+	 * @since 0.5.0
+	 * @param {*} value The value to parse
+	 * @param {SchemaPiece} schemaPiece The SchemaPiece which manages this value
+	 * @returns {*}
+	 * @private
+	 * @static
+	 */
+	static _parseSQLValue(value, schemaPiece) {
+		if (typeof value !== 'undefined') {
+			if (schemaPiece.array) {
+				if (value === null) return schemaPiece.default.slice(0);
+				if (typeof value === 'string') value = tryParse(value);
+				if (Array.isArray(value)) value = value.map(val => GatewayStorage._parseSQLValue(val, schemaPiece));
+				return value;
+			}
+			if (schemaPiece.type === 'any') {
```

Comment:
Switchcase would probably be better here instead of if else if else if

Reply:
Oh yeah, could be. Next commit comes with a cleanup for this method 👍